### PR TITLE
compute-client: add an explicit drop flag for collections

### PR DESCRIPTION
**Pull request by teskje**
So far the compute controller considered a collection dropped when its read frontier has advanced to the empty frontier. This had the defect that we might remove a collection's state from the controller in response to a client setting a `ReadPolicy::ValidFrom([])`. A client might still expect to be able to query information about a collection even when it is not interested in reading data from it anymore.

To remove this defect, a `CollectionState::dropped` flag is introduced and used to explicitly mark collections as dropped when `drop_collections` is called. Only collections that have the `dropped` flag set are then considered for having their controller state dropped.

Figuring out the right places in which to check the collection state removal conditions becomes a bit of a hassle with this change, so we also rearrange things so that the collection state cleanup is performed asynchronously in a period task. This makes us keep unneeded collections around a bit longer than necessary but also greatly simplifies the rest of the code.

### Motivation

  * This PR fixes a recognized bug.

I split this out from the fix for https://github.com/MaterializeInc/materialize/issues/24469 because said fix turns out to be more complex than I initially expected.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/A

### CODE REVIEW

**Review by ggevay — APPROVED** [review]
I have some minor comments about some comments, but otherwise LGTM. It very much makes sense to me to make collection dropping more explicit than just as an implicit response to frontier changes.

**Comment by antiguru** [comment]
> A client might still expect to be able to query information about a collection even when it is not interested in reading data from it anymore.

Can you elaborate what information about a collection some client will expect after advancing the read frontier to `[]`?

File: src/compute-client/src/controller/instance.rs [hunk 1]
```diff
@@ -303,6 +303,40 @@ impl<T> Instance<T> {
         self.deliver_introspection_updates(IntrospectionType::ComputeDependencies, updates);
     }
 
+    /// Clean up collection state that is not needed anymore.
+    ///
+    /// Three conditions need to be true before we can remove state for a collection:
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/controller/instance.rs:301]: All fixed. Thanks for being so thorough!

**Comment by teskje** [comment]
> Can you elaborate what information about a collection some client will expect after advancing the read frontier to []?

Our controller API is underspecified, but I would expect this sequence of calls to work:

```
controller.create_dataflow('x', description);
controller.set_read_policy('x', ReadPolicy::ValidFrom(Antichain::new()));
let coll = controller.collection('x');
// query some metadata from `coll`, like frontiers or dependencies
controller.drop_collections(['x']); 
```

In today's implementation the last two controller invocations might fail because we've already dropped the collection when the read policy was set. That makes for a confusing API, even if you could maybe argue that there is no reason to want to know about collection metadata when you don't want to read from the collection anymore (I'm not sure about that).

Apart from API considerations, the reason we need this is that we want to downgrade read holds for write-only collections (like MVs) more aggressively. When an MV advances to the empty frontier, we can also drop all read holds we have taken for the MV, which with the current behavior would let the controller drop the collection state too. That wouldn't work well with adapter which expects calls for an MV to succeed until it is dropped.

**Review by vmarcos — APPROVED** [review]
LGTM! (ref discussion in [Slack](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1705919765555909))

**Comment by teskje** [comment]
TFTRs!

**Pull request merged**